### PR TITLE
[CARBONDATA-3444]Fix MV query failure when column name and table name is same in case of join scenario

### PR DESCRIPTION
**Pull request by akashrn5**
### Problem:
when there are columns with same in different table, after sql generation, the project column will be like `gen_subsumer_0.product` , it fails during logical plan generation from rewritten query, as column names will be ambigous

### Solution:
update the outputlist when there are duplicate columns present in query. Here we can form the qualified name for the Attribute reference. So when qualifier is defined for column, the qualified name wil be like 
`<col_qualifier_name>_<col.name>`, if qualifier is not defined, then it will be `<col_exprId_id>_<col.name>` . So update for all the nodes like groupby , select nodes, so that it will be handled when there will be amguity in columns.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 NA
 - [x] Any backward compatibility impacted?
 NA
 - [x] Document update required?
NA
 - [x] Testing done
Added UT
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

File: datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/util/Logical2ModularExtractions.scala [hunk 1]
```diff
@@ -146,6 +148,37 @@ object ExtractSelectModule extends PredicateHelper {
...
+  def updateOutputListOfJoin(output: Seq[Attribute], condition: Seq[Expression]): Seq[Attribute] = {
+    val fieldsInCondition = condition.collect {
+      case a: Expression => a.children.collect {
+        case a: AttributeReference =>
+          a.toAttribute
+      }
+    }.flatten
+    if (condition.nonEmpty && fieldsInCondition.nonEmpty) {
+      val substitutedOutputs: Seq[Attribute] = output.map {
+        col =>
```
Comment (ravipesala) [inline datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/util/Logical2ModularExtractions.scala]: move up

**Comment by ravipesala** [comment]
LGTM, apart from minor comment.

**Comment by akashrn5** [comment]
@ravipesala handled comments, please review

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3835/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11902/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3629/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11909/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3635/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3842/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3846/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3646/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11920/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3856/

**Comment by akashrn5** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3654/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3864/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11928/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3659/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3868/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11932/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3661/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11934/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3870/

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala [hunk 2]
```diff
@@ -124,7 +124,14 @@ object SelectSelectNoChildDelta extends DefaultMatchPattern with PredicateHelper
           exprListR.exists(a1 => a1.isInstanceOf[Alias] &&
                                  a1.asInstanceOf[Alias].child.semanticEquals(a.child)) ||
           exprListR.exists(_.semanticEquals(exprE) || canEvaluate(exprE, subsumer))
-        case exp => exprListR.exists(_.semanticEquals(exp) || canEvaluate(exp, subsumer))
+        case exp =>
+          if (exprListR.collect { case a@Alias(_, _) => a }.nonEmpty) {
```
Comment (ravipesala) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala]: Why is this `if` condition is required? 
you can directly write 
```
 case exp =>
exprListR.exists(a1 => a1.isInstanceOf[Alias] &&
                            a1.asInstanceOf[Alias].child.semanticEquals(exp)) ||
exprListR.exists(_.semanticEquals(exp) || canEvaluate(exp, subsumer))
```

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala [hunk 3]
```diff
@@ -1092,6 +1092,18 @@ class MVCreateTestCase extends QueryTest with BeforeAndAfterAll {
     assert(TestUtil.verifyMVDataMap(analyzed4, "constant_mv"))
   }
 
+  test("test mv query when the column names and table name same in join scenario") {
```
Comment (ravipesala) [inline datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: Add some more tests with join , groupby with creating mv with both table duplicate columns , with and with out aliases

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala [hunk 4]
```diff
@@ -310,4 +310,35 @@ object MVUtil {
...
     }
   }
+
+  def updateDuplicateColumns(outputList: Seq[NamedExpression]): Seq[NamedExpression] = {
+    val duplicateNameCols = outputList.groupBy(_.name).filter(_._2.length > 1).flatMap(_._2)
+      .toList
+    val updatedOutList = outputList.map { col =>
+      val duplicateColumn = duplicateNameCols
+        .find(a => a.semanticEquals(col))
+      val qualifiedName = col.qualifier.getOrElse(s"${ col.name }_${ col.exprId.id }")
```
Comment (ravipesala) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala]: I think it supposed to be 
```
val qualifiedName = col.name +"_" +col.qualifier.getOrElse(s"${ col.exprId.id }")
```

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala [hunk 5]
```diff
@@ -310,4 +310,35 @@ object MVUtil {
...
+      if (duplicateColumn.isDefined) {
+        val attributesOfDuplicateCol = duplicateColumn.get.collect {
+          case a: AttributeReference => a
+        }
+        val attributeOfCol = col.collect { case a: AttributeReference => a }
+        val isStrictDuplicate = attributesOfDuplicateCol.forall(expr =>
+          attributeOfCol.exists(a => a.semanticEquals(expr)))
+        if (!isStrictDuplicate) {
```
Comment (ravipesala) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala]: This `isStrictDuplicate` part is not clear the need of it, please add some comments to it.

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala [hunk 6]
```diff
@@ -310,4 +310,35 @@ object MVUtil {
...
+          Alias(col, qualifiedName)(exprId = col.exprId)
+        } else if (col.qualifier.isDefined) {
+          Alias(col, col.qualifiedName)(exprId = col.exprId)
+        } else if (duplicateColumn.get.isInstanceOf[AttributeReference] &&
```
Comment (ravipesala) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala]: I am not sure this `if` condition is needed

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala (hunk 3 above)
Comment (akashrn5) [reply datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: done

**Comment by akashrn5** [comment]
@ravipesala handled review comments, and given comment for some checks, please hep review

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala (hunk 3 above)
Comment (akashrn5) [reply datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVCreateTestCase.scala]: done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3662/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11935/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3871/

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala (hunk 5 above)
Comment (akashrn5) [reply datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala]: i have added the comments, done

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala (hunk 6 above)
Comment (akashrn5) [reply datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVUtil.scala]: i have added the comments with scenario for this check, please help review

**Comment by ravipesala** [comment]
LGTM
